File: test/config.test.ts
import { describe, it, expect, beforeEach, afterEach, spyOn } from "bun:test";
import path from "path";

const TEST_DIR = path.join(import.meta.dir, "..", ".test-tmp");

async function cleanup() {
  try {
    await Bun.spawn(["rm", "-rf", TEST_DIR]).exited;
  } catch {}
}



describe("Config", () => {
  beforeEach(async () => {
    await cleanup();
    const config = await import("../src/config");
    config.invalidateConfigCache(TEST_DIR);
  });

  afterEach(async () => {
    await cleanup();
  });

  it("detects Drupal root from docroot/core", async () => {
    await Bun.spawn(["mkdir", "-p", path.join(TEST_DIR, "docroot", "core")]).exited;
    const { detectDrupalRoot } = await import("../src/config");
    expect(detectDrupalRoot(TEST_DIR)).toBe("docroot");
  });

  it("detects Drupal root from web/core", async () => {
    await Bun.spawn(["mkdir", "-p", path.join(TEST_DIR, "web", "core")]).exited;
    const { detectDrupalRoot } = await import("../src/config");
    expect(detectDrupalRoot(TEST_DIR)).toBe("web");
  });

  it("creates default config on first load", async () => {
    const { loadConfig } = await import("../src/config");
    const config = await loadConfig(TEST_DIR);
    expect(config.routes).toContain("docroot/modules/custom");
    expect(config.routes).toContain("docroot/themes/custom");
    expect(config.debounce).toBe(800);
    expect(config.drushCommand).toBe("cr");
    expect(config.patterns).toContain(".php");
  });

  it("loads existing config file", async () => {
    await Bun.spawn(["mkdir", "-p", path.join(TEST_DIR, "docroot", "core")]).exited;
    const customConfig = {
      routes: ["docroot/modules/custom", "docroot/themes/custom", "docroot/modules/contrib"],
      patterns: [".html.twig", ".module", ".theme"],
      debounce: 600,
      drushCommand: "cr",
    };
    await Bun.write(path.join(TEST_DIR, "watcher.config.json"), JSON.stringify(customConfig, null, 2));

    const { loadConfig, invalidateConfigCache } = await import("../src/config");
    invalidateConfigCache(TEST_DIR);
    const config = await loadConfig(TEST_DIR);
    expect(config.routes).toHaveLength(3);
    expect(config.debounce).toBe(600);
  });

  it("returns null for no Drupal root", async () => {
    const { detectDrupalRoot } = await import("../src/config");
    expect(detectDrupalRoot(TEST_DIR)).toBeNull();
  });
});

describe("Drush", () => {
  it("returns the configured drushCmd when set", async () => {
    const { getDrushCommand } = await import("../src/drush");
    const cmd = getDrushCommand({ drushCmd: "/custom/drush" });
    expect(cmd).toBe("/custom/drush");
  });

  it("returns 'drush' as fallback", async () => {
    const { getDrushCommand } = await import("../src/drush");
    const cmd = getDrushCommand({});
    expect(cmd).toBe("drush");
  });

  it("builds spawn args from config", async () => {
    const { getDrushSpawnArgs } = await import("../src/drush");
    const { cmd, args } = getDrushSpawnArgs({ drushCmd: "ddev drush", drushCommand: "cr" });
    expect(cmd).toBe("ddev");
    expect(args).toEqual(["drush", "cr"]);
  });

  it("includes extra drush args", async () => {
    const { getDrushSpawnArgs } = await import("../src/drush");
    const { cmd, args } = getDrushSpawnArgs({
      drushCmd: "drush",
      drushCommand: "cr",
      drushArgs: ["--uri=default"],
    });
    expect(cmd).toBe("drush");
    expect(args).toEqual(["cr", "--uri=default"]);
  });
});

describe("Main", () => {
  it("parses --abort-on-drush-error flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--abort-on-drush-error"]);
    expect(flags.abortOnDrushError).toBe(true);
  });

  it("parses --watch flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--watch=modules/custom"]);
    expect(flags.watchRoutes).toEqual(["modules/custom"]);
  });

  it("parses --no-watch flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--no-watch=modules/contrib"]);
    expect(flags.noWatchRoutes).toEqual(["modules/contrib"]);
  });

  it("parses multiple flags at once", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags([
      "--abort-on-drush-error",
      "--watch=modules/custom",
      "--no-watch=modules/contrib",
    ]);
    expect(flags.abortOnDrushError).toBe(true);
    expect(flags.watchRoutes).toEqual(["modules/custom"]);
    expect(flags.noWatchRoutes).toEqual(["modules/contrib"]);
  });

  it("handles empty argv", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags, extra } = parseFlags([]);
    expect(flags.abortOnDrushError).toBe(false);
    expect(flags.watchRoutes).toEqual([]);
    expect(flags.noWatchRoutes).toEqual([]);
    expect(extra).toEqual([]);
  });

  it("parses --dry-run flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--dry-run"]);
    expect(flags.dryRun).toBe(true);
  });

  it("parses --verbose flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--verbose"]);
    expect(flags.verbose).toBe(true);
  });

  it("parses -v alias", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["-v"]);
    expect(flags.verbose).toBe(true);
  });

  it("parses --no-colors flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--no-colors"]);
    expect(flags.noColors).toBe(true);
  });

  it("combines all new flags", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--dry-run", "--verbose", "--no-colors"]);
    expect(flags.dryRun).toBe(true);
    expect(flags.verbose).toBe(true);
    expect(flags.noColors).toBe(true);
  });

  it("parses --debounce flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--debounce=1200"]);
    expect(flags.debounce).toBe(1200);
  });

  it("parses --no-dotfiles flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--no-dotfiles"]);
    expect(flags.noDotfiles).toBe(true);
  });

  it("parses --log-file flag", async () => {
    const { parseFlags } = await import("../src/main");
    const { flags } = parseFlags(["--log-file=/tmp/watcher.log"]);
    expect(flags.logFile).toBe("/tmp/watcher.log");
  });
});

describe("Drush duration", () => {
  it("includes duration in runDrush result", async () => {
    const { runDrush } = await import("../src/drush");
    const result = await runDrush("echo", ["hello"]);
    expect(result).toHaveProperty("exitCode");
    expect(result).toHaveProperty("duration");
    expect(typeof result.duration).toBe("string");
    expect(Number(result.duration)).toBeGreaterThanOrEqual(0);
  });
});

describe("Colors", () => {
  it("setColorsEnabled toggles colors", async () => {
    const utils = await import("../src/utils");
    utils.setColorsEnabled(false);
    expect(utils.colorsEnabled()).toBe(false);
    expect(utils.green("test")).toBe("test");
    expect(utils.red("test")).toBe("test");
    utils.setColorsEnabled(true);
    expect(utils.colorsEnabled()).toBe(true);
    expect(utils.green("test")).not.toBe("test");
  });
});

describe("Config validation", () => {
  it("routes must be an array", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({ routes: null }, TEST_DIR);
    expect(Array.isArray(config.routes)).toBe(true);
    expect(config.routes.length).toBeGreaterThan(0);
  });

  it("debounce must be a positive number", async () => {
    const { validateConfig } = await import("../src/config");
    const c1 = validateConfig({ debounce: "abc" }, TEST_DIR);
    expect(c1.debounce).toBe(800);
    const c2 = validateConfig({ debounce: -100 }, TEST_DIR);
    expect(c2.debounce).toBe(800);
  });

  it("drushArgs must be an array", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({ drushArgs: "not-an-array" }, TEST_DIR);
    expect(Array.isArray(config.drushArgs)).toBe(true);
  });

  it("preserves valid values", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({
      routes: ["docroot/modules/custom"],
      debounce: 500,
      drushArgs: ["--uri=default"],
    }, TEST_DIR);
    expect(config.routes).toEqual(["docroot/modules/custom"]);
    expect(config.debounce).toBe(500);
    expect(config.drushArgs).toEqual(["--uri=default"]);
  });

  it("accepts valid commandsPerPattern merged with defaults", async () => {
    const { validateConfig } = await import("../src/config");
    const cpp = { ".html.twig": "custom-command" };
    const config = validateConfig({ commandsPerPattern: cpp }, TEST_DIR);
    // User override takes precedence
    expect(config.commandsPerPattern[".html.twig"]).toBe("custom-command");
    // Defaults still present
    expect(config.commandsPerPattern[".theme"]).toBe("cc theme-registry");
    expect(config.commandsPerPattern[".php"]).toBe("cc plugin");
    expect(config.commandsPerPattern[".yml"]).toBe("cc plugin");
  });

  it("fills defaults from old config file with empty commandsPerPattern", async () => {
    // Simulate a config file created before commandsPerPattern defaults existed
    const oldConfig = { routes: ["docroot/modules/custom"], commandsPerPattern: {} };
    await Bun.write(path.join(TEST_DIR, "watcher.config.json"), JSON.stringify(oldConfig));
    const { loadConfig, invalidateConfigCache } = await import("../src/config");
    invalidateConfigCache(TEST_DIR);
    const config = await loadConfig(TEST_DIR);
    // In-memory config should have the full defaults merged
    expect(config.commandsPerPattern[".html.twig"]).toBe("cc twig");
    expect(config.commandsPerPattern[".php"]).toBe("cc plugin");
    expect(config.commandsPerPattern[".yml"]).toBe("cc plugin");
    // File on disk still has the old value (no overwrite)
    const onDisk = JSON.parse(await Bun.file(path.join(TEST_DIR, "watcher.config.json")).text());
    expect(onDisk.commandsPerPattern).toEqual({});
  });

  it("uses all defaults when commandsPerPattern is empty", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({ commandsPerPattern: {} }, TEST_DIR);
    expect(config.commandsPerPattern[".html.twig"]).toBe("cc twig");
    expect(config.commandsPerPattern[".module"]).toBe("cc plugin");
  });

  it("rejects non-object commandsPerPattern", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({ commandsPerPattern: "invalid" }, TEST_DIR);
    expect(typeof config.commandsPerPattern).toBe("object");
    expect(Array.isArray(config.commandsPerPattern)).toBe(false);
    // Falls back to all defaults
    expect(config.commandsPerPattern[".html.twig"]).toBe("cc twig");
    expect(config.commandsPerPattern[".php"]).toBe("cc plugin");
  });

  it("getCacheClearArgs returns drush cr when no commandsPerPattern", async () => {
    const mod = await import("../src/watcher");
    // Private function — test via the export that uses it indirectly
    // Instead, verify that getDrushSpawnArgs behavior is preserved
    const { getDrushSpawnArgs } = await import("../src/drush");
    const { cmd, args } = getDrushSpawnArgs({ drushCommand: "cr" });
    expect(cmd).toBe("drush");
    expect(args).toContain("cr");
  });

  it("normalizes routes (removes trailing slashes)", async () => {
    const { validateConfig } = await import("../src/config");
    const config = validateConfig({ routes: ["docroot/modules/custom/", "docroot/themes//custom"] }, TEST_DIR);
    expect(config.routes[0]).toBe("docroot/modules/custom");
    expect(config.routes[1]).toBe("docroot/themes/custom");
  });

  it("validates invalid config when loading", async () => {
    await Bun.write(
      path.join(TEST_DIR, "watcher.config.json"),
      JSON.stringify({ routes: null, debounce: "abc" }, null, 2)
    );
    const { loadConfig, invalidateConfigCache } = await import("../src/config");
    invalidateConfigCache(TEST_DIR);
    const config = await loadConfig(TEST_DIR);
    expect(Array.isArray(config.routes)).toBe(true);
    expect(typeof config.debounce).toBe("number");
  });
});

describe("PID check", () => {
  it("returns null when no PID file", async () => {
    const { checkPid } = await import("../src/config");
    expect(await checkPid(TEST_DIR)).toBeNull();
  });

  it("returns stale for non-existent PID", async () => {
    const { checkPid } = await import("../src/config");
    const pidPath = path.join(TEST_DIR, ".drupal-watcher.pid");
    await Bun.write(pidPath, "99999999");
    const result = await checkPid(TEST_DIR);
    expect(result).toBe("stale");
  });

  it("returns null for empty PID file", async () => {
    const { checkPid } = await import("../src/config");
    const pidPath = path.join(TEST_DIR, ".drupal-watcher.pid");
    await Bun.write(pidPath, "");
    const result = await checkPid(TEST_DIR);
    expect(result).toBeNull();
  });

  it("starttime round-trip works", async () => {
    const { writeStarttime, getStarttime, removeStarttime } = await import("../src/config");
    await writeStarttime(TEST_DIR);
    const t = await getStarttime(TEST_DIR);
    expect(typeof t).toBe("number");
    expect(t).toBeGreaterThan(0);
    expect(Date.now() - t).toBeLessThan(5000);
    await removeStarttime(TEST_DIR);
    expect(await getStarttime(TEST_DIR)).toBeNull();
  });
});

describe("Watcher", () => {
  it("resetDebounce clears state", async () => {
    const { resetDebounce, stats } = await import("../src/watcher");
    resetDebounce();
    expect(stats.changes).toBe(0);
    expect(stats.clears).toBe(0);
    expect(stats.startTime).toBeNull();
    expect(stats.filesChanged instanceof Set).toBe(true);
  });
});

describe("Commands", () => {
  it("cmdHelp prints general help", async () => {
    const spy = spyOn(console, "log");
    const { cmdHelp } = await import("../src/commands");
    cmdHelp();
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it("cmdHelp prints start help", async () => {
    const spy = spyOn(console, "log");
    const { cmdHelp } = await import("../src/commands");
    cmdHelp("start");
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it("cmdStatus shows not running when no PID", async () => {
    const spy = spyOn(console, "log");
    const { cmdStatus } = await import("../src/commands");
    await cmdStatus();
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  it("cmdRestart does not throw when no watcher", async () => {
    const spy = spyOn(console, "log");
    const { cmdRestart } = await import("../src/commands");
    // This will call cmdStart but fail early at health check since no Drupal root
    // We just verify the restart function itself is callable
    expect(typeof cmdRestart).toBe("function");
    spy.mockRestore();
  });
});

describe("Drush shell", () => {
  it("healthCheck returns false on invalid drush", async () => {
    const { healthCheck } = await import("../src/drush");
    const result = await healthCheck({ drushCmd: "/nonexistent/drush" });
    expect(result).toBe(false);
  });

  it("uses sh -c on non-Windows", async () => {
    const { runPostClearCommands } = await import("../src/drush");
    // Spawn and immediately cancel to verify it doesn't throw
    const cmd = "echo test-shell";
    const proc = Bun.spawn(["sh", "-c", cmd]);
    const exitCode = await proc.exited;
    expect(exitCode).toBe(0);
  });
});

File: internal/config/config_test.go
package config_test

import (
	"os"
	"path/filepath"
	"testing"

	"github.com/irving-frias/drupal-watcher/internal/config"
)

func TestNewManager(t *testing.T) {
	m := config.NewManager()
	if m == nil {
		t.Fatal("expected non-nil manager")
	}
}

func TestDetectDrupalRoot(t *testing.T) {
	tmp := t.TempDir()
	m := config.NewManager()

	// No Drupal root
	root := m.DetectDrupalRoot(tmp)
	if root != nil {
		t.Errorf("expected nil, got %v", *root)
	}

	// Create a Drupal-like structure
	docroot := filepath.Join(tmp, "docroot")
	if err := os.MkdirAll(filepath.Join(docroot, "core"), 0755); err != nil {
		t.Fatal(err)
	}

	root = m.DetectDrupalRoot(tmp)
	if root == nil || *root != "docroot" {
		t.Errorf("expected docroot, got %v", root)
	}

	// Test with web directory
	tmp2 := t.TempDir()
	web := filepath.Join(tmp2, "web")
	if err := os.MkdirAll(filepath.Join(web, "core"), 0755); err != nil {
		t.Fatal(err)
	}

	m2 := config.NewManager()
	root = m2.DetectDrupalRoot(tmp2)
	if root == nil || *root != "web" {
		t.Errorf("expected web, got %v", root)
	}
}

func TestGetDefaultConfig(t *testing.T) {
	m := config.NewManager()
	cfg := m.GetDefaultConfig("")

	if len(cfg.Routes) == 0 {
		t.Error("expected routes")
	}
	if cfg.Debounce != 800 {
		t.Errorf("expected debounce 800, got %d", cfg.Debounce)
	}
	if cfg.DrushCommand != "cr" {
		t.Errorf("expected drushCommand cr, got %s", cfg.DrushCommand)
	}
	if cfg.CommandsPerPattern == nil {
		t.Error("expected commandsPerPattern")
	}
}

func TestValidateConfig(t *testing.T) {
	m := config.NewManager()

	// Empty config should get defaults
	cfg := m.ValidateConfig(config.Config{}, "")
	if len(cfg.Routes) == 0 {
		t.Error("expected routes after validation")
	}
	if cfg.Debounce <= 0 {
		t.Errorf("expected positive debounce, got %d", cfg.Debounce)
	}

	// Custom values should be preserved (with defaults merged)
	customCfg := config.Config{
		Routes:   []string{"custom/route"},
		Debounce: 500,
	}
	cfg = m.ValidateConfig(customCfg, "")
	if len(cfg.Routes) != 1 || cfg.Routes[0] != "custom/route" {
		t.Errorf("expected custom route, got %v", cfg.Routes)
	}
	if cfg.Debounce != 500 {
		t.Errorf("expected debounce 500, got %d", cfg.Debounce)
	}
	if cfg.DrushCommand == "" {
		t.Error("expected default drushCommand")
	}
}

func TestSaveAndLoadConfig(t *testing.T) {
	tmp := t.TempDir()
	m := config.NewManager()

	cfg := config.Config{
		Routes:   []string{"test/routes"},
		Debounce: 300,
		DrushCommand: "cc all",
	}

	if err := m.SaveConfig(cfg, tmp); err != nil {
		t.Fatal(err)
	}

	loaded, err := m.LoadConfig(tmp)
	if err != nil {
		t.Fatal(err)
	}

	if len(loaded.Routes) != 1 || loaded.Routes[0] != "test/routes" {
		t.Errorf("expected test/routes, got %v", loaded.Routes)
	}
	if loaded.Debounce != 300 {
		t.Errorf("expected 300, got %d", loaded.Debounce)
	}
}

func TestLoadConfigInvalidJSON(t *testing.T) {
	tmp := t.TempDir()
	m := config.NewManager()

	os.WriteFile(filepath.Join(tmp, "watcher.config.json"), []byte("{invalid"), 0644)

	// Should fall back to defaults without error
	cfg, err := m.LoadConfig(tmp)
	if err != nil {
		t.Fatal(err)
	}
	if len(cfg.Routes) == 0 {
		t.Error("expected defaults for invalid JSON")
	}
}

func TestCache(t *testing.T) {
	tmp := t.TempDir()
	m := config.NewManager()

	// Load once
	cfg1, _ := m.LoadConfig(tmp)
	// Load again (should use cache)
	cfg2, _ := m.LoadConfig(tmp)

	if &cfg1 == &cfg2 {
		t.Error("LoadConfig should return a copy, not cached pointer")
	}

	// Invalidate
	m.InvalidateConfigCache(tmp)
	cfg3, _ := m.LoadConfig(tmp)
	if cfg3.Debounce != cfg1.Debounce {
		t.Errorf("expected same debounce after reload, got %d vs %d", cfg3.Debounce, cfg1.Debounce)
	}
}

func TestPIDWriteAndCheck(t *testing.T) {
	tmp := t.TempDir()

	if err := config.WritePid(tmp); err != nil {
		t.Fatal(err)
	}

	status, err := config.CheckPid(tmp)
	if err != nil {
		t.Fatal(err)
	}
	if status == nil {
		t.Fatal("expected PID to exist")
	}
	if status == "stale" {
		t.Fatal("expected valid PID")
	}
}

func TestPIDRemove(t *testing.T) {
	tmp := t.TempDir()

	config.WritePid(tmp)
	config.RemovePid(tmp)

	status, err := config.CheckPid(tmp)
	if err != nil {
		t.Fatal(err)
	}
	if status != nil {
		t.Errorf("expected nil after removal, got %v", status)
	}
}

func TestStarttime(t *testing.T) {
	tmp := t.TempDir()

	if err := config.WriteStarttime(tmp); err != nil {
		t.Fatal(err)
	}

	st, err := config.GetStarttime(tmp)
	if err != nil {
		t.Fatal(err)
	}
	if st <= 0 {
		t.Errorf("expected positive starttime, got %d", st)
	}

	config.RemoveStarttime(tmp)
	st, err = config.GetStarttime(tmp)
	if err != nil {
		t.Fatal(err)
	}
	if st != 0 {
		t.Errorf("expected 0 after removal, got %d", st)
	}
}

func TestCommandsPerPatternMerge(t *testing.T) {
	m := config.NewManager()
	cfg := config.Config{
		CommandsPerPattern: map[string]string{
			".php": "custom command",
		},
	}
	validated := m.ValidateConfig(cfg, "")

	if validated.CommandsPerPattern[".html.twig"] != "cc twig" {
		t.Error("expected default for .html.twig to be preserved")
	}
	if validated.CommandsPerPattern[".php"] != "custom command" {
		t.Errorf("expected custom override, got %s", validated.CommandsPerPattern[".php"])
	}
}

func TestRouteNormalization(t *testing.T) {
	m := config.NewManager()
	cfg := config.Config{
		Routes: []string{"./docroot//modules/custom/"},
	}
	validated := m.ValidateConfig(cfg, "")

	if len(validated.Routes) != 1 {
		t.Fatalf("expected 1 route, got %v", validated.Routes)
	}
	expected := "docroot/modules/custom"
	if validated.Routes[0] != expected {
		t.Errorf("expected %s, got %s", expected, validated.Routes[0])
	}
}

func TestGetPidFile(t *testing.T) {
	path := config.GetPidFile("/test/root")
	if !filepath.IsAbs(path) && path != "/test/root/.drupal-watcher.pid" {
		// GetPidFile just delegates to pidPath
	}
}
